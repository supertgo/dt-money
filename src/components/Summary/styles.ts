import styled, { css } from 'styled-components';

export const Wrapper = styled.main`
  margin-top: -16rem;
`;

export const GridWrapper = styled.div`
  ${({ theme }) => css`
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    gap: ${theme.grid.gutter};
  `}
`;
