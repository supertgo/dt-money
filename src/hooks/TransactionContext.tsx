import { createContext, useState, useEffect, useContext } from 'react';
import { api } from 'services/api';
import { getStorageItem, setStorageItem } from 'utils/localStorage';

const TRANSACTION_kEY = 'transactionsItems';

export type TransactionProps = {
  id: number;
  title: string;
  type: string;
  amount: number;
  category: string;
  createdAt: string;
};

type TransactionInput = Omit<TransactionProps, 'id' | 'createdAt'>;

type TransactionsProviderProps = {
  children: React.ReactNode;
};

type TransactionContextProps = {
  transactions: TransactionProps[];
  createTransaction: (transaction: TransactionInput) => Promise<void>;
};

const TransactionContext = createContext<TransactionContextProps>(
  {} as TransactionContextProps
);

export const TransactionsProvider = ({
  children
}: TransactionsProviderProps) => {
  const [transactions, setTransactions] = useState<TransactionProps[]>([]);

  useEffect(() => {
    const data = getStorageItem(TRANSACTION_kEY);

    if (data) {
      setTransactions(data);
    }
  }, []);

  async function createTransaction(transactionInput: TransactionInput) {
    const response = await api.post('/transactions', {
      ...transactionInput,
      createdAt: new Date()
    });
    const { transaction } = response.data;

    saveTransactions([...transactions, transaction]);
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function saveTransactions(transaction: any) {
    setStorageItem(TRANSACTION_kEY, transaction);
    setTransactions(transaction);
  }

  return (
    <TransactionContext.Provider value={{ transactions, createTransaction }}>
      {children}
    </TransactionContext.Provider>
  );
};

export const useTransactions = () => useContext(TransactionContext);
