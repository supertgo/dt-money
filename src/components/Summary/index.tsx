import MediaMatch from 'components/MediaMatch';
import TransactionBox from 'components/TransactionBox';
import TransactionBoxSlider from 'components/TransactionBoxSlider';
import { useTransactions } from 'hooks/TransactionContext';

import * as S from './styles';

const Summary = () => {
  const { transactions } = useTransactions();

  const summary = transactions.reduce(
    (acc, transaction) => {
      if (transaction.type === 'deposit') {
        acc.deposits += transaction.amount;
        acc.total += transaction.amount;
      } else {
        acc.withdraws += transaction.amount;
        acc.total -= transaction.amount;
      }

      return acc;
    },
    {
      deposits: 0,
      withdraws: 0,
      total: 0
    }
  );

  return (
    <S.Wrapper>
      <MediaMatch lessThan="medium">
        <TransactionBoxSlider>
          <TransactionBox amount={summary.deposits} />
          <TransactionBox type="outcome" amount={summary.withdraws} />
          <TransactionBox type="total" amount={summary.total} color="green" />
        </TransactionBoxSlider>
      </MediaMatch>

      <MediaMatch greaterThan="medium">
        <S.GridWrapper>
          <TransactionBox amount={summary.deposits} />
          <TransactionBox type="outcome" amount={summary.withdraws} />
          <TransactionBox type="total" amount={summary.total} color="green" />
        </S.GridWrapper>
      </MediaMatch>
    </S.Wrapper>
  );
};

export default Summary;
