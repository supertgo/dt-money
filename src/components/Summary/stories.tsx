import { Story, Meta } from '@storybook/react';
import { TransactionsProvider } from 'hooks/TransactionContext';
import Summary from '.';

export default {
  title: 'Summary',
  component: Summary
} as Meta;

export const Default: Story = () => (
  <TransactionsProvider>
    <Summary />
  </TransactionsProvider>
);
