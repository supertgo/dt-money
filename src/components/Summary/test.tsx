import 'match-media-mock';

import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helpers';
import { TransactionsProvider } from 'hooks/TransactionContext';

import Summary from '.';

jest.mock('components/TransactionBox', () => ({
  __esModule: true,
  default: function Mock() {
    return <div data-testid="Mock TransactionBox" />;
  }
}));

describe('<Summary />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(
      <TransactionsProvider>
        <Summary />
      </TransactionsProvider>
    );

    expect(screen.getAllByTestId('Mock TransactionBox')).toHaveLength(6);
    expect(container).toMatchSnapshot();
  });
});
