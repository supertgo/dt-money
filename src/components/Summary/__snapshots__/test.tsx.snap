// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Summary /> should render the heading 1`] = `
.c1 {
  display: none;
}

.c4 {
  display: none;
}

.c3 .slick-slider {
  position: relative;
  display: block;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-touch-callout: none;
  -khtml-user-select: none;
  -ms-touch-action: pan-y;
  touch-action: pan-y;
  -webkit-tap-highlight-color: transparent;
}

.c3 .slick-list {
  position: relative;
  display: block;
  overflow: hidden;
  margin: 0;
  padding: 0;
}

.c3 .slick-list:focus {
  outline: none;
}

.c3 .slick-list.dragging {
  cursor: pointer;
  cursor: hand;
}

.c3 .slick-slider .slick-track,
.c3 .slick-slider .slick-list {
  -webkit-transform: translate3d(0,0,0);
  -moz-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  -o-transform: translate3d(0,0,0);
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
}

.c3 .slick-track {
  position: relative;
  top: 0;
  left: 0;
  display: block;
}

.c3 .slick-track:before,
.c3 .slick-track:after {
  display: table;
  content: '';
}

.c3 .slick-track:after {
  clear: both;
}

.c3 .slick-loading .slick-track {
  visibility: hidden;
}

.c3 .slick-slide {
  display: none;
  float: left;
  height: 100%;
  min-height: 1px;
}

.c3 [dir='rtl'] .slick-slide {
  float: right;
}

.c3 .slick-slide img {
  display: block;
}

.c3 .slick-slide.slick-loading img {
  display: none;
}

.c3 .slick-slide.dragging img {
  pointer-events: none;
}

.c3 .slick-initialized .slick-slide {
  display: block;
}

.c3 .slick-loading .slick-slide {
  visibility: hidden;
}

.c3 .slick-vertical .slick-slide {
  display: block;
  height: auto;
  border: 1px solid transparent;
}

.c3 .slick-arrow.slick-hidden {
  display: none;
}

.c2 .slick-track {
  width: 1000rem !important;
}

.c2 .slick-slide {
  margin-left: 1rem;
  width: 25rem !important;
}

.c0 {
  margin-top: -16rem;
}

.c5 {
  display: grid;
  grid-template-columns: repeat(3,1fr);
  gap: 3.2rem;
}

@media (max-width:768px) {
  .c1 {
    display: block;
  }
}

@media (min-width:768px) {
  .c4 {
    display: block;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
    >
      <main
        class="c2"
      >
        <main
          class="c3"
        >
          <div
            class="slick-slider slick-initialized"
            dir="ltr"
          >
            <div
              class="slick-list"
            >
              <div
                class="slick-track"
                style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-hidden="false"
                  class="slick-slide slick-active slick-current"
                  data-index="0"
                  style="outline: none; width: 0px;"
                  tabindex="-1"
                >
                  <div>
                    <div
                      data-testid="Mock TransactionBox"
                    />
                  </div>
                </div>
                <div
                  aria-hidden="true"
                  class="slick-slide"
                  data-index="1"
                  style="outline: none; width: 0px;"
                  tabindex="-1"
                >
                  <div>
                    <div
                      data-testid="Mock TransactionBox"
                    />
                  </div>
                </div>
                <div
                  aria-hidden="true"
                  class="slick-slide"
                  data-index="2"
                  style="outline: none; width: 0px;"
                  tabindex="-1"
                >
                  <div>
                    <div
                      data-testid="Mock TransactionBox"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </main>
      </main>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          data-testid="Mock TransactionBox"
        />
        <div
          data-testid="Mock TransactionBox"
        />
        <div
          data-testid="Mock TransactionBox"
        />
      </div>
    </div>
  </main>
</div>
`;
